pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/database/IdentifiersDaoTest.scala

```diff
@@ -99,55 +169,84 @@ class IdentifiersDaoTest
+            _ <- identifiersDao.saveIdentifiers(
+              List(duplicatedIdentifier1, duplicatedIdentifier2))
+            insertResult <- identifiersDao.saveIdentifiers(identifiers)
+          } yield (insertResult)
+
+          result shouldBe a[Failure[_]]
+          val error = result.failed.get.asInstanceOf[IdentifiersDao.InsertError]
+          error.e shouldBe a[BatchUpdateException]
+          error.failed should contain theSameElementsAs (identifiers
+            .filter(i =>
+              i.CanonicalId == duplicatedIdentifier1.CanonicalId || i.CanonicalId == duplicatedIdentifier2.CanonicalId))
+          error.succeeded should contain theSameElementsAs (identifiers
+            .filterNot(i =>
+              i.CanonicalId == duplicatedIdentifier1.CanonicalId || i.CanonicalId == duplicatedIdentifier2.CanonicalId))
+
+          val lookupResult = identifiersDao.lookupIds(ids.keys.toList).get
+          lookupResult.found.values should contain theSameElementsAs (error.succeeded)
       }
     }
 
     it(
-      "saves records with the same SourceSystem and SourceId but different OntologyType") {
-      val sourceIdentifier1 = createSourceIdentifierWith(
-        ontologyType = "Foo"
-      )
-      val sourceIdentifier2 = createSourceIdentifierWith(
-        ontologyType = "Bar"
-      )
-
-      val identifier1 = createSQLIdentifierWith(
-        sourceIdentifier = sourceIdentifier1
-      )
-
-      val identifier2 = createSQLIdentifierWith(
-        sourceIdentifier = sourceIdentifier2
-      )
+      "fails saving the identifiers if the same OntologyType,SourceName, and SourceId are already present") {
+      val identifier = createSQLIdentifier
+      val duplicateIdentifier = identifier.copy(CanonicalId = createCanonicalId)
 
       withIdentifiersDao {
         case (identifiersDao, _) =>
-          identifiersDao.saveIdentifier(identifier1) shouldBe Success(1)
-          identifiersDao.saveIdentifier(identifier2) shouldBe Success(1)
+          val result = for {
+            _ <- identifiersDao.saveIdentifiers(List(identifier))
+            insertResult <- identifiersDao.saveIdentifiers(
+              List(duplicateIdentifier))
+          } yield (insertResult)
+
+          result shouldBe a[Failure[_]]
+          val error = result.failed.get.asInstanceOf[IdentifiersDao.InsertError]
+          error.e shouldBe a[BatchUpdateException]
+          error.failed shouldBe List(duplicateIdentifier)
       }
     }
 
     it(
-      "saves records with different SourceId but the same OntologyType and SourceSystem") {
-      val sourceIdentifier1 = createSourceIdentifierWith(
-        value = "1234"
-      )
-      val sourceIdentifier2 = createSourceIdentifierWith(
-        value = "5678"
-      )
+      "saves records with the same SourceSystem and SourceId but different OntologyType") {
+      val sourceIdentifier1 = createSourceIdentifier
+      val sourceIdentifier2 = sourceIdentifier1.copy(ontologyType = "Foo")
```

Comment:
An ontologyType of `"Foo"` is probably less likely from `createSourceIdentifier`, but would still be good to be explicit that they're different.